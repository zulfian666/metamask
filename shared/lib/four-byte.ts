import { MethodRegistry } from 'eth-method-registry';
import fetchWithCache from './fetch-with-cache';

type FourByteResult = {
  created_at: string;
  text_signature: string;
};

type FourByteResponse = {
  results: FourByteResult[];
};

export async function getMethodFrom4Byte(
  fourBytePrefix: string,
): Promise<string> {
  const fourByteResponse = (await fetchWithCache({
    url: `https://www.4byte.directory/api/v1/signatures/?hex_signature=${fourBytePrefix}`,
    fetchOptions: {
      referrerPolicy: 'no-referrer-when-downgrade',
      body: null,
      method: 'GET',
      mode: 'cors',
    },
    functionName: 'getMethodFrom4Byte',
  })) as FourByteResponse;

  fourByteResponse.results.sort((a, b) => {
    return new Date(a.created_at).getTime() < new Date(b.created_at).getTime()
      ? -1
      : 1;
  });

  return fourByteResponse.results[0].text_signature;
}

let registry: MethodRegistry | undefined;

type HttpProvider = {
  host: string;
  timeout: number;
};

type MethodRegistryArgs = {
  network: string;
  provider: HttpProvider;
};

export async function getMethodDataAsync(
  fourBytePrefix: string,
  allow4ByteRequests: boolean,
  provider?: unknown,
) {
  try {
    let fourByteSig = null;
    if (allow4ByteRequests) {
      fourByteSig = await getMethodFrom4Byte(fourBytePrefix).catch((e) => {
        console.error(e);
        return null;
      });
      console.log('fourByteSig = ', fourByteSig);
    }

    if (!registry) {
      registry = new MethodRegistry({
        provider: provider ?? global.ethereumProvider,
      } as MethodRegistryArgs);
    }

    if (!fourByteSig) {
      return {};
    }

    const parsedResult = registry.parse(fourByteSig);

    return {
      name: parsedResult.name,
      params: parsedResult.args,
    };
  } catch (error) {
    console.error(error);
    return {};
  }
}
