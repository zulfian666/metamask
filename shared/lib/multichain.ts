import { isEthAddress } from '../../app/scripts/lib/multichain/address';

/**
 * Returns whether an address is on the Bitcoin mainnet.
 *
 * This function only checks the prefix of the address to determine if it's on
 * the mainnet or not. It doesn't validate the address itself, and should only
 * be used as a temporary solution until this information is included in the
 * account object.
 *
 * @param address - The address to check.
 * @returns `true` if the address is on the Bitcoin mainnet, `false` otherwise.
 */
export function isBtcMainnetAddress(address: string): boolean {
  return (
    !isEthAddress(address) &&
    (address.startsWith('bc1') || address.startsWith('1'))
  );
}

/**
 * Returns whether an address is on the Bitcoin testnet.
 *
 * See {@link isBtcMainnetAddress} for implementation details.
 *
 * @param address - The address to check.
 * @returns `true` if the address is on the Bitcoin testnet, `false` otherwise.
 */
export function isBtcTestnetAddress(address: string): boolean {
  return !isEthAddress(address) && !isBtcMainnetAddress(address);
}
