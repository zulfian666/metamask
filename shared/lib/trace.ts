import * as Sentry from '@sentry/browser';
import { Primitive, StartSpanOptions } from '@sentry/types';
import { createModuleLogger } from '@metamask/utils';
import { log as sentryLogger } from '../../app/scripts/lib/setupSentry';

export enum TraceName {
  BackgroundConnect = 'Background Connect',
  DeveloperTest = 'Developer Test',
  FirstRender = 'First Render',
  GetState = 'Get State',
  InitialActions = 'Initial Actions',
  LoadScripts = 'Load Scripts',
  Middleware = 'Middleware',
  NestedTest1 = 'Nested Test 1',
  NestedTest2 = 'Nested Test 2',
  NotificationDisplay = 'Notification Display',
  PPOMValidation = 'PPOM Validation',
  SetupStore = 'Setup Store',
  Transaction = 'Transaction',
  UIStartup = 'UI Startup',
}

const log = createModuleLogger(sentryLogger, 'trace');

const ID_DEFAULT = 'default';
const OP_DEFAULT = 'custom';

const tracesByKey: Map<string, PendingTrace> = new Map();

type PendingTrace = {
  end: (timestamp?: number) => void;
  request: TraceRequest;
  startTime: number;
};

export type TraceContext = unknown;

export type TraceCallback<T> = (context?: TraceContext) => T;

export type TraceRequest = {
  data?: Record<string, number | string | boolean>;
  id?: string;
  name: TraceName;
  parentContext?: TraceContext;
  startTime?: number;
  tags?: Record<string, number | string | boolean>;
};

export type EndTraceRequest = {
  id?: string;
  name: TraceName;
  timestamp?: number;
};

export function trace<T>(request: TraceRequest, fn: TraceCallback<T>): T;

export function trace(request: TraceRequest): TraceContext;

export function trace<T>(
  request: TraceRequest,
  fn?: TraceCallback<T>,
): T | TraceContext {
  if (!fn) {
    return startTrace(request);
  }

  return traceCallback(request, fn);
}

export function endTrace(request: EndTraceRequest) {
  const { name, timestamp } = request;
  const id = getTraceId(request);
  const key = getTraceKey(request);
  const pendingTrace = tracesByKey.get(key);

  if (!pendingTrace) {
    log('No pending trace found', name, id);
    return;
  }

  pendingTrace.end(timestamp);

  tracesByKey.delete(key);

  const { request: pendingRequest, startTime } = pendingTrace;
  const endTime = timestamp ?? getPerformanceTimestamp();
  const duration = endTime - startTime;

  log('Finished trace', name, id, duration, { request: pendingRequest });
}

function traceCallback<T>(request: TraceRequest, fn: TraceCallback<T>): T {
  const { name } = request;

  const callback = (span: Sentry.Span | null) => {
    log('Starting trace', name, request);

    const start = Date.now();
    let error: unknown;

    return tryCatchMaybePromise<T>(
      () => fn(span),
      (currentError) => {
        error = currentError;
        throw currentError;
      },
      () => {
        const end = Date.now();
        const duration = end - start;

        log('Finished trace', name, duration, { error, request });
      },
    ) as T;
  };

  return startSpan(request, (spanOptions) =>
    Sentry.startSpan(spanOptions, callback),
  );
}

function startTrace(request: TraceRequest): TraceContext {
  const { name, startTime: requestStartTime } = request;
  const startTime = requestStartTime ?? getPerformanceTimestamp();
  const id = getTraceId(request);

  const callback = (span: Sentry.Span | null) => {
    const end = (timestamp?: number) => {
      span?.end(timestamp);
    };

    const pendingTrace = { end, request, startTime };
    const key = getTraceKey(request);
    tracesByKey.set(key, pendingTrace);

    log('Started trace', name, id, request);

    return span;
  };

  return startSpan(request, (spanOptions) =>
    Sentry.startSpanManual(spanOptions, callback),
  );
}

function startSpan<T>(
  request: TraceRequest,
  callback: (spanOptions: StartSpanOptions) => T,
) {
  const { data: attributes, name, parentContext, startTime, tags } = request;
  const parentSpan = (parentContext ?? null) as Sentry.Span | null;

  const spanOptions: StartSpanOptions = {
    attributes,
    name,
    op: OP_DEFAULT,
    parentSpan,
    startTime,
  };

  return Sentry.withIsolationScope((scope) => {
    scope.setTags(tags as Record<string, Primitive>);

    return callback(spanOptions);
  });
}

function getTraceId(request: TraceRequest) {
  return request.id ?? ID_DEFAULT;
}

function getTraceKey(request: TraceRequest) {
  const { name } = request;
  const id = getTraceId(request);

  return [name, id].join(':');
}

function getPerformanceTimestamp(): number {
  return performance.timeOrigin + performance.now();
}

function tryCatchMaybePromise<T>(
  tryFn: () => T,
  catchFn: (error: unknown) => void,
  finallyFn: () => void,
): T | undefined {
  let isPromise = false;

  try {
    const result = tryFn() as T;

    if (result instanceof Promise) {
      isPromise = true;
      return result.catch(catchFn).finally(finallyFn) as T;
    }

    return result;
  } catch (error) {
    if (!isPromise) {
      catchFn(error);
    }
  } finally {
    if (!isPromise) {
      finallyFn();
    }
  }

  return undefined;
}
