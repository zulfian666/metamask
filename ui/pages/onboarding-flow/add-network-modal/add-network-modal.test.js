import React from 'react';
import { waitFor } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { renderWithProvider } from '../../../../test/lib/render-helpers';
import AddNetworkModal from '.';

const mockHideModal = jest.fn();
jest.mock('../../../store/actions', () => ({
  ...jest.requireActual('../../../store/actions'),
  hideModal: () => mockHideModal,
}));

const mockNetworkMenuRedesignToggle = jest.fn();

jest.mock('../../../helpers/utils/feature-flags', () => ({
  ...jest.requireActual('../../../helpers/utils/feature-flags'),
  getLocalNetworkMenuRedesignFeatureFlag: () => mockNetworkMenuRedesignToggle,
}));

describe('Add Network Modal', () => {
  it('should render', async () => {
    mockNetworkMenuRedesignToggle.mockImplementation(() => false);

    const mockStore = configureMockStore([])({
      metamask: { useSafeChainsListValidation: true, orderedNetworkList: {} },
    });

    const { container } = renderWithProvider(
      <AddNetworkModal showHeader />,
      mockStore,
    );

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('should not render the new network flow modal', async () => {
    mockNetworkMenuRedesignToggle.mockReturnValue(true);

    const mockStore = configureMockStore([thunk])({
      metamask: { useSafeChainsListValidation: true, orderedNetworkList: {} },
    });

    const { queryByText } = renderWithProvider(
      <AddNetworkModal showHeader />,
      mockStore,
    );

    await waitFor(() => {
      expect(queryByText('Cancel')).not.toBeInTheDocument();
      expect(queryByText('Next')).toBeInTheDocument();
    });
  });
});
