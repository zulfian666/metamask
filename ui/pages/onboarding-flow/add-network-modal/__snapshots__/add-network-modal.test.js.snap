// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Network Modal should render 1`] = `
<div>
  <div
    class="box box--padding-top-4 box--flex-direction-row"
  >
    <h4
      class="box box--margin-top-1 box--margin-bottom-1 box--flex-direction-row typography typography--h4 typography--weight-bold typography--style-normal typography--align-center typography--color-text-default"
    >
      Add custom network
    </h4>
  </div>
  <div
    class="mm-box networks-tab__scrollable mm-box--padding-top-1 mm-box--padding-right-4 mm-box--padding-bottom-2 mm-box--padding-left-4 mm-box--display-flex mm-box--flex-direction-column mm-box--align-items-center"
  >
    <div
      class="networks-tab__add-network-form networks-tab__restrict-height"
    >
      <div
        class="actionable-message actionable-message--warning actionable-message--with-right-button networks-tab__add-network-form__alert actionable-message--with-icon"
      >
        <svg
          viewBox="0 0 10 10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 0C2.2 0 0 2.2 0 5s2.2 5 5 5 5-2.2 5-5-2.2-5-5-5zm0 2c.4 0 .7.3.7.7s-.3.7-.7.7-.7-.2-.7-.6.3-.8.7-.8zm.7 6H4.3V4.3h1.5V8z"
            fill="var(--color-warning-default)"
          />
        </svg>
        
        <div
          class="actionable-message__message"
        >
          A malicious network provider can lie about the state of the blockchain and record your network activity. Only add custom networks you trust.
        </div>
      </div>
      <div
        class="networks-tab__add-network-form-body"
      >
        <div
          class="box mm-form-text-field box--padding-top-4 box--padding-bottom-4 box--display-flex box--flex-direction-column"
          data-testid="network-form-name-input"
        >
          <label
            class="mm-box mm-text mm-label mm-form-text-field__label mm-text--body-sm mm-text--font-weight-bold mm-box--padding-top-1 mm-box--padding-bottom-1 mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
          >
            Network name
          </label>
          <div
            class="mm-box mm-text-field mm-text-field--size-md mm-text-field--truncate mm-form-text-field__text-field mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--align-items-center mm-box--background-color-background-default mm-box--rounded-sm mm-box--border-width-1 box--border-style-solid"
          >
            <input
              autocomplete="off"
              class="mm-box mm-text mm-input mm-input--disable-state-styles mm-text-field__input mm-text--body-sm mm-box--margin-0 mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-2 mm-box--color-text-default mm-box--background-color-transparent mm-box--border-style-none"
              data-testid="network-form-network-name"
              focused="false"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="form-field"
        >
          <label
            class="mm-box"
          >
            <div
              class="form-field__heading"
            >
              <div
                class="mm-box form-field__heading-title mm-box--display-flex mm-box--align-items-baseline"
              >
                <h6
                  class="mm-box mm-text mm-text--body-sm-bold mm-box--display-inline-block mm-box--color-text-default"
                >
                  New RPC URL
                </h6>
                
                
              </div>
              
            </div>
            <input
              class="form-field__input"
              data-testid="network-form-rpc-url"
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          class="box mm-form-text-field box--padding-top-4 box--display-flex box--flex-direction-column"
          data-testid="network-form-chain-id-input"
        >
          <label
            class="mm-box mm-text mm-label mm-form-text-field__label mm-text--body-sm mm-text--font-weight-bold mm-box--padding-top-1 mm-box--padding-bottom-1 mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
          >
            Chain ID
          </label>
          <div
            class="mm-box mm-text-field mm-text-field--size-md mm-text-field--truncate mm-form-text-field__text-field mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--align-items-center mm-box--background-color-background-default mm-box--rounded-sm mm-box--border-width-1 box--border-style-solid"
          >
            <input
              autocomplete="off"
              class="mm-box mm-text mm-input mm-input--disable-state-styles mm-text-field__input mm-text--body-sm mm-box--margin-0 mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-2 mm-box--color-text-default mm-box--background-color-transparent mm-box--border-style-none"
              data-testid="network-form-chain-id"
              focused="false"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="box mm-form-text-field box--padding-top-4 box--display-flex box--flex-direction-column"
          data-testid="network-form-ticker"
        >
          <label
            class="mm-box mm-text mm-label mm-form-text-field__label mm-text--body-sm mm-text--font-weight-bold mm-box--padding-top-1 mm-box--padding-bottom-1 mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
          >
            Currency symbol
          </label>
          <div
            class="mm-box mm-text-field mm-text-field--size-md mm-text-field--truncate mm-form-text-field__text-field mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--align-items-center mm-box--background-color-background-default mm-box--rounded-sm mm-box--border-width-1 box--border-style-solid"
          >
            <input
              autocomplete="off"
              class="mm-box mm-text mm-input mm-input--disable-state-styles mm-text-field__input mm-text--body-sm mm-box--margin-0 mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-2 mm-box--color-text-default mm-box--background-color-transparent mm-box--border-style-none"
              data-testid="network-form-ticker-input"
              focused="false"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="box mm-form-text-field box--padding-top-4 box--display-flex box--flex-direction-column"
          data-testid="network-form-block-explorer-url-input"
        >
          <label
            class="mm-box mm-text mm-label mm-form-text-field__label mm-text--body-sm mm-text--font-weight-bold mm-box--padding-top-1 mm-box--padding-bottom-1 mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
          >
            Block explorer URL (Optional)
          </label>
          <div
            class="mm-box mm-text-field mm-text-field--size-md mm-text-field--truncate mm-form-text-field__text-field mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--align-items-center mm-box--background-color-background-default mm-box--rounded-sm mm-box--border-width-1 box--border-style-solid"
          >
            <input
              autocomplete="off"
              class="mm-box mm-text mm-input mm-input--disable-state-styles mm-text-field__input mm-text--body-sm mm-box--margin-0 mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-2 mm-box--color-text-default mm-box--background-color-transparent mm-box--border-style-none"
              data-testid="network-form-block-explorer-url"
              focused="false"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="mm-box mm-box--padding-top-4 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--text-align-center mm-box--width-full mm-box--background-color-background-default"
    >
      <button
        class="mm-box mm-text mm-button-base mm-button-base--size-lg mm-button-base--disabled mm-button-primary mm-button-primary--disabled mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--width-full mm-box--color-primary-inverse mm-box--background-color-primary-default mm-box--rounded-pill"
        disabled=""
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
